src/blockrelay/blockrelay_common.cpp

```diff
@@ -210,6 +210,18 @@ void ThinTypeRelay::ClearBlockInFlight(CNode *pfrom, const uint256 &hash)
     }
 }
 
+void ThinTypeRelay::ClearAllBlocksInFlight(NodeId id)
+{
+    LOCK(cs_inflight);
+    std::pair<std::multimap<const NodeId, CThinTypeBlockInFlight>::iterator,
+        std::multimap<const NodeId, CThinTypeBlockInFlight>::iterator>
+        range = mapThinTypeBlocksInFlight.equal_range(id);
+    while (range.first != range.second)
```

Comment:
https://en.cppreference.com/w/cpp/container/map/equal_range:
```If there are no elements not less than key, past-the-end (see end()) iterator is returned as the first element. Similarly if there are no elements greater than key, past-the-end iterator is returned as the second element.```

In the case where you try to find an id that is lower than everything in the map, I read this as returning (end(), iter_to_first_elem).

This code would then execute mapThinTypeBlocksInFlight.erase(end()), which is illegal based on the spec of std::map::erase: ```The iterator pos must be valid and dereferenceable. Thus the end() iterator (which is valid, but is not dereferenceable) cannot be used as a value for pos.```

But the fix pretty easy -- just add a check before the while:
if (range.first != mapThinTypeBlocksInFlight.end()) while (range.first != range.second)
{
...
}

Reply:
If no matches are found both elements point to the end() iterator and so the loop will never run because range.first will equal range.second.

Reply:
why would they both point to the end()?  The spec says that given the set 1,2,3, a search for 0 would return (end(), 1)

Reply:
NVM, misread "no elements not less" as "no elements less"